SegmentEditorCropSegment/SegmentEditorCropSegmentLib/SegmentEditorEffect.py

```diff
@@ -0,0 +1,419 @@
+      self.outputVolumeSelector.setCurrentNode(outputVolume)
+
+    slicer.app.setOverrideCursor(qt.Qt.WaitCursor)
+    padExtent = [self.xPad.value, self.zPad.value, self.yPad.value]
+    self.cropVolumeWithSegment(segmentationNode, segmentID, inputVolume, outputVolume, padExtent, self.fillValue.value)
+    qt.QApplication.restoreOverrideCursor()
+ 
+ 
+  def cropVolumeWithSegment(self, segmentationNode, segmentID, inputVolumeNode, outputVolumeNode, padRegion, padValue):
+    """
+    Fill voxels of the input volume inside/outside the Cropping model with the provided fill value
+    """
+ 
+    segmentIDs = vtk.vtkStringArray()
+    segmentIDs.InsertNextValue(segmentID)
+    intensityRange = inputVolumeNode.GetImageData().GetScalarRange()
+    maskedVolume = slicer.modules.volumes.logic().CreateAndAddLabelVolume(inputVolumeNode, "TemporaryVolumeCrop")
+    if not maskedVolume:
+      logging.error("cropVolumeWithSegment failed: invalid maskedVolume")
+      return False
+ 
+    if not slicer.vtkSlicerSegmentationsModuleLogic.ExportSegmentsToLabelmapNode(segmentationNode, segmentIDs, maskedVolume, inputVolumeNode):
+      logging.error("cropVolumeWithSegment failed: ExportSegmentsToLabelmapNode error")
+      slicer.mrmlScene.RemoveNode(maskedVolume.GetDisplayNode().GetColorNode())
+      slicer.mrmlScene.RemoveNode(maskedVolume.GetDisplayNode())
+      slicer.mrmlScene.RemoveNode(maskedVolume)
+      return False
+ 
+    # Crop to region greater than threshold
+    maskToStencil = vtk.vtkImageToImageStencil()
+    maskToStencil.ThresholdByLower(0)
+    maskToStencil.SetInputData(maskedVolume.GetImageData())
+    stencil = vtk.vtkImageStencil()
+    stencil.SetInputData(inputVolumeNode.GetImageData())
+    stencil.SetStencilConnection(maskToStencil.GetOutputPort())
+    stencil.SetReverseStencil(1)
+    stencil.SetBackgroundValue(intensityRange[0]-1)
+    stencil.Update()
+     
+    outputVolumeNode.SetAndObserveImageData(stencil.GetOutput())
+    # Set the same geometry and parent transform as the input volume
+    ijkToRas = vtk.vtkMatrix4x4()
+    inputVolumeNode.GetIJKToRASMatrix(ijkToRas)
+    outputVolumeNode.SetIJKToRASMatrix(ijkToRas)
+    inputVolumeNode.SetAndObserveTransformNodeID(inputVolumeNode.GetTransformNodeID())
+     
+    # Get masked output as a vtk oriented image and crop
+    cropThreshold = 0
+    img = slicer.modules.segmentations.logic().CreateOrientedImageDataFromVolumeNode(outputVolumeNode) 
+    img.UnRegister(None) 
+    extent=[0,0,0,0,0,0]
+    vtkSegmentationCore.vtkOrientedImageDataResample.CalculateEffectiveExtent(img, extent, cropThreshold) 
+    croppedImg = vtkSegmentationCore.vtkOrientedImageData() 
+    vtkSegmentationCore.vtkOrientedImageDataResample.CopyImage(img, croppedImg, extent) 
+    slicer.modules.segmentations.logic().CopyOrientedImageDataToVolumeNode(croppedImg, outputVolumeNode)
+    
+    # Use ITK to apply pad to image
+    itkImage = sitkUtils.PullVolumeFromSlicer(outputVolumeNode) 

```

Comment:
Thanks for this feedback @lassoan . Could you clarify how I can access maskVolumeWithSegment? I can get the mask volume effect using: 

> effect = slicer.qMRMLSegmentEditorWidget().effectByName('Mask volume')

but can't access maskVolumeWithSegment(). Is there a better way to do this? Thanks for your advice.

Reply:
`slicer.qMRMLSegmentEditorWidget()` creates a new segment editor widget and it gets deleted immediately. You probably want to retrieve the effect from the current segment editor widget instead of creating a new one.

Instead, you could retrieve a valid instance of the effect from Segment Editor module's widget:

```
effect=slicer.modules.segmenteditor.widgetRepresentation().self().editor.effectByName('Mask volume').self()
effect.maskVolumeWithSegment(...)
```

(`self()` retrieves Python implementation of a class that has a C++ interface)

As you can see, retrieving this utility method from an existing is somewhat too complicated, so I would recommend a nicer solution: make `maskVolumeWithSegment` method a static method like this:

```
  @staticmethod
  def maskVolumeWithSegment(segmentationNode, segmentID, operationMode, fillValues, inputVolumeNode, outputVolumeNode):
```

This static method could then be simply called as:

```
import SegmentEditorMaskVolumeLib
SegmentEditorMaskVolumeLib.SegmentEditorEffect.maskVolumeWithSegment(…)
```

Reply:
Thanks @lassoan, this makes sense and the second method is working for me.